HSBM.py

```diff
@@ -0,0 +1,391 @@
+            cluster = AutoGMMCluster(
+                min_components=1, max_components=self.max_components, **self.cluster_kws
+            )
+            pred = cluster.fit_predict(X)
+            model = cluster.model_
+            bic = model.bic(X)
+            lik = model.score(X)
+            k = cluster.n_components_
+
+        elif self.cluster_method == "Kmeans":
+            cluster = KMeansCluster(
+                max_clusters=self.max_components, **self.cluster_kws
+            )
+            cluster.fit(X)
+            model = cluster.model_
+            k = cluster.n_clusters_
+            pred = cluster.predict(X)
+
+        results = {
+            "bic": -bic,
+            "lik": lik,
+            "k": k,
+            "model": model,
+            "pred": pred,
+        }
+        self.results_ = results
+
+    def _select_model(self, k=None):
+        model = self.results_["model"]
+        pred = self.results_["pred"]
+        k = self.results_["k"]
+
+        self.k_ = k
+        self.children = []
+
+        if k > 1:
+            self.model_ = model
+            self.pred_ = pred
+            root_labels = self.root.labels
+
+            pred_name = f"{self.depth + 1}_pred"
+            if pred_name not in root_labels.columns:
+                root_labels[pred_name] = ""
+            root_labels.loc[self.root_inds, pred_name] = pred
+            uni_labels = np.unique(pred)
+
+            for ul in uni_labels:
+                new_labels = root_labels[
+                    (root_labels[pred_name] == ul)
+                    & (root_labels.index.isin(self.root_inds.index))
+                ]
+                new_root_inds = new_labels["inds"]
+                new_x = self.root.X[new_root_inds]
+
+                if self.adj is not None:
+                    new_adj = self.root.adj[
+                        np.ix_(new_root_inds, new_root_inds)
+                    ]
+                    if self.reembed is True:
+                        new_x = self._embed(new_adj, reembedding=True)
+
+                RecursiveCluster(
+                    selection_criteria_GMM=self.selection_criteria_GMM,
+                    cluster_method=self.cluster_method,
+                    root_inds=new_root_inds,
+                    n_init=self.n_init,
+                    parent=self,
+                    max_components=self.max_components,
+                    min_split=self.min_split,
+                    cluster_kws=self.cluster_kws,
+                    reembed=self.reembed,
+                )
+                self.children[-1].X = new_x
+
+                if self.adj is not None:
+                    self.children[-1].adj = new_adj
+                else:
+                    self.children[-1].adj = self.adj
+            
+    # TODO: maybe add tuning params like delta_bic or likelihood ratio
+    # for model selection
+
+    def _embed(self, graph, reembedding=False):
+        raise NotImplementedError()
+
+    def _collect_labels(self):
```

Comment:
i think we need to get rid of this function, was very hacky on my part, and replace with something recursive.

Reply:
note that this is kinda specific to the input data